### PR TITLE
ENH, BLD: Add configure option for blas/lapack libraries

### PR DESCRIPTION
**Pull request by Smit-create**
We can now switch blas/lapack libraries (after `meson setup`) in the following way:

```
$ cd build
$ meson configure # list all the options
$ meson configure -Dblas=mkl  # Change blas library
```
cc @rgommers

### CODE REVIEW

**Comment by ev-br** [comment]
1) how does one select a system-provided blas/lapack? 

2) what is meson_options.txt, do we edit it manually or it's autogenerated? is it possible to fold it all into meson.build?

**Comment by Smit-create** [comment]
> 1. how does one select a system-provided blas/lapack?

By default it uses `openblas` for both blas and lapack, but you can change to any other library using `meson configure -Dblas=mkl -Dlapack=mkl`.

> 2. what is meson_options.txt, do we edit it manually or it's autogenerated? is it possible to fold it all into meson.build?

It is not autogenerated. I have added it manually for meson options and is editable. The use of `meson_options.txt` is documented here: https://mesonbuild.com/Build-options.html#build-options

**Comment by ev-br** [comment]
Tried this PR on a machine which does not have openblas.

```
$ meson setup build --prefix=$PWD/installdir
The Meson build system
Version: 0.61.1
Source dir: /home/br/repos/scipy/scipy
Build dir: /home/br/repos/scipy/scipy/build
Build type: native build
Project name: SciPy
Project version: 1.9.0.dev0
C compiler for the host machine: cc (gcc 9.3.0 "cc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0")
C linker for the host machine: cc ld.bfd 2.34
C++ compiler for the host machine: c++ (gcc 9.3.0 "c++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0")
C++ linker for the host machine: c++ ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Compiler for C supports arguments -Wno-unused-but-set-variable: YES 
Library m found: YES
Fortran compiler for the host machine: gfortran (gcc 9.3.0 "GNU Fortran (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0")
Fortran linker for the host machine: gfortran ld.bfd 2.34
Program cython found: YES (/home/br/.virtualenvs/scipy_meson/bin/cython)
Program pythran found: YES (/home/br/.virtualenvs/scipy_meson/bin/pythran)
Program cp found: YES (/usr/bin/cp)
Program python3 found: YES (/home/br/.virtualenvs/scipy_meson/bin/python)
WARNING: Python files installed by Meson might not be found by python interpreter.
 This warning can be avoided by setting "python.platlibdir" option.
WARNING: Python files installed by Meson might not be found by python interpreter.
 This warning can be avoided by setting "python.purelibdir" option.
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Library npymath found: YES
Library npyrandom found: YES
Did not find CMake 'cmake'
Found CMake: NO
Run-time dependency openblas found: NO (tried pkgconfig)

scipy/meson.build:111:0: ERROR: Dependency "openblas" not found, tried pkgconfig

A full log can be found at /home/br/repos/scipy/scipy/build/meson-logs/meson-log.txt
(scipy_meson) br@gonzales:~/repos/scipy/scipy$ cd build/
(scipy_meson) br@gonzales:~/repos/scipy/scipy/build$ meson configure

ERROR: No such build data file as "'/home/br/repos/scipy/scipy/build/meson-private/build.dat'".
(scipy_meson) br@gonzales:~/repos/scipy/scipy/build$ cd ..
(scipy_meson) br@gonzales:~/repos/scipy/scipy$ meson configure
WARNING: The source directory instead of the build directory was specified.
WARNING: Only the default values for the project are printed, and all command line parameters are ignored.
```

Running `meson configure` in the `build` directory just fails; running it from the scipy root gives a list of options, however the blas part is not very helpful:

```
  Project options    Default Value        Possible Values      Description
  ---------------    -------------        ---------------      -----------
  blas               openblas                                  option for BLAS library
                                                               switching
  lapack             openblas                                  option for LAPACK library
                                                               switching
```

So the original question still stands: what are available blas/lapack options, and how to select them? You only mention MKL. MKL is great and all, other options are not implemented yet, or need some syntax?

**Comment by ilayn** [comment]
Indeed this is often the most confusing part of building scipy and we can't allow for half-working options. Plus if possible we have to print where it actually found the library. Otherwise all kinds of weird stuff is happening and this is not helping the user. 

In other words, the "error message design"-wise, it doesn't tell what happened when also doesn't offer a possible solution or available options. The last part can be omitted but the first part can't.

**Comment by ev-br** [comment]
Looking into meson-log.txt, it might be that one other issue is at play :

```
Pkg-config binary for 1 is cached.
Determining dependency 'openblas' with pkg-config executable '/usr/bin/pkg-config'
env[PKG_CONFIG_PATH]: 
Called `/usr/bin/pkg-config --modversion openblas` -> 1

CMake binary for 1 is not cached
CMake binary missing from cross or native file, or env var undefined.
Trying a default CMake fallback at cmake
Did not find CMake 'cmake'
Found CMake: NO
CMake binary for machine 1 not found. Giving up.
Run-time dependency openblas found: NO (tried pkgconfig)

scipy/meson.build:111:0: ERROR: Dependency "openblas" not found, tried pkgconfig
```

I'm pretty sure `setup.py`-based build worked without cmake, re-running it now to double-check. But I don't know if it's relevant or not.

**Comment by Smit-create** [comment]
> Running `meson configure` in the `build` directory just fails; running it from the scipy root gives a list of options, however the blas part is not very helpful:

This works fine on my machine:
```
$ meson configure
...
  Project options       Default Value           Possible Values         Description
  ---------------       -------------           ---------------         -----------
  blas                  openblas                                        option for BLAS library switching
  lapack                openblas                                        option for LAPACK library switching
$ meson configure -Dblas=mkl
$ meson configure
...
  Project options       Current Value           Possible Values         Description
  ---------------       -------------           ---------------         -----------
  blas                  mkl                                             option for BLAS library switching
  lapack                openblas                                        option for LAPACK library switching

```

**Comment by ilayn** [comment]
Indeed a standalone MKL doesn't go through but with OpenBLAS it selects MKL

**Comment by rgommers** [comment]
So a bit of context:
- There was a TODO about correctly incorporating the `g77` ABI wrappers, this PR does that
- It also adds the user-level config option we need to select different BLAS/LAPACK libraries (so it's the equivalent of `site.cfg`)
- Currently that means that BLAS and LAPACK libraries that will work solely using `pkg-config` (like Netlib) should work. Apparently there's an issue in `odr/meson.build` to fix for this.
- MKL will not work yet, because its pkg-config support is completely broken and hence it needs custom support within Meson itself so the discovery and compile/link flags are correct.

**Comment by ilayn** [comment]
Can pkg-config be a bit verbose? Currently it is not very helpful where it searched or what it searched etc. I am constantly struggling with this in setup.cfg. I don't understand why people don't add this very simple log item option in these complicated tools.

**Comment by ev-br** [comment]
> Currently that means that BLAS and LAPACK libraries that will work solely using pkg-config (like Netlib) should work.

and 

> The undefined symbol: cspmv_ issue can be looked into, that should be fixable without too much trouble.

I'm ready to test with a netlib-only setup once it's ready for testing :-).

**Comment by rgommers** [comment]
> Can pkg-config be a bit verbose? Currently it is not very helpful where it searched or what it searched etc.

The log in `build/meson-logs/meson-log.txt` has:
```
Determining dependency 'openblas' with pkg-config executable '/home/rgommers/anaconda3/envs/scipy-dev/bin/pkg-config'
env[PKG_CONFIG_PATH]: 
Called `/home/rgommers/anaconda3/envs/scipy-dev/bin/pkg-config --modversion openblas` -> 0
0.3.17
env[PKG_CONFIG_PATH]: 
Called `/home/rgommers/anaconda3/envs/scipy-dev/bin/pkg-config --cflags openblas` -> 0
-I/home/rgommers/anaconda3/envs/scipy-dev/include
env[PKG_CONFIG_ALLOW_SYSTEM_LIBS]: 1
env[PKG_CONFIG_PATH]: 
Called `/home/rgommers/anaconda3/envs/scipy-dev/bin/pkg-config --libs openblas` -> 0
-L/home/rgommers/anaconda3/envs/scipy-dev/lib -lopenblas
env[PKG_CONFIG_PATH]: 
Called `/home/rgommers/anaconda3/envs/scipy-dev/bin/pkg-config --libs openblas` -> 0
-L/home/rgommers/anaconda3/envs/scipy-dev/lib -lopenblas
```
That seems pretty detailed; that kind of output is not shown by default which is probably a good thing. When things fail it should show something like that in the stdout logging (can't verify right now).

**Comment by ilayn** [comment]
Ah then it is my bad. I didn't think of checking the log file. I thought all was pasted here. Nevermind then

**Comment by eli-schwartz** [comment]
> 2\. what is meson_options.txt, do we edit it manually or it's autogenerated? is it possible to fold it all into meson.build?

Meson design goals have decided to make project options easily discoverable, so they must be defined in a separate file which is always at the root of the project. Granted that some project authors want to have "one less file", there are nevertheless good reasons for it which won't change.

Benefits:
- Users do not have to hunt through miles worth of build descriptions to find the names and meanings of options.
- Users don't have to ask whether an option exists or not, depending on whether the option was defined inside of an `if` block
- Meson doesn't have to parse anything other than that one options-only file in order to pretty-print the available options.



> Running `meson configure` in the `build` directory just fails

That is unfortunately the only logical outcome configuring the build directory fails. :( It's essentially unusable, so we don't use it.

Running inside the source directory should be fine though -- it's equivalent to autotools `./configure --help` and you simply don't get to take advantage of the enhanced ability to list currently active values.

> I'm pretty sure `setup.py`-based build worked without cmake, re-running it now to double-check. But I don't know if it's relevant or not.

This is not relevant. To clarify what is going on here: Meson was told to unconditionally look up a dependency (by the name of "openblas") and it tried to do so using its standard lookup methods:
- pkg-config
- cmake
- optionally some stuff only relevant if defined in Meson's source code ("custom lookup functionality")

When pkg-config failed, it tried cmake. But cmake did not exist (it wasn't even installed) so as a result it disabled that lookup method.

If cmake worked, *and* there was an `openblas-config.cmake`, it could try running a trace on a specially crafted CMakeLists.txt to see if the crafted cmake project could find an "openblas" dependency. This would fail anyway: openblas itself provides a pkg-config dependency called `openblas` and also a cmake dependency called `OpenBLAS`, and they are case sensitive so you'd need to look for both. But also, pkg-config will never fail in a case where cmake succeeds.

The main point here is that cmake is a last ditch "Hail Mary" attempt to find a `dependency('openblas')`, it isn't actually required. Note that when Meson failed, it didn't say "error: couldn't find cmake", it said "error: couldn't find openblas using pkg-config".

> That seems pretty detailed; that kind of output is not shown by default which is probably a good thing. When things fail it should show something like that in the stdout logging (can't verify right now).

Meson doesn't show the same detailed debug logging when it errors. It does, however, tell you the file path that "a full log" can be found in. IMO this is the right balance to make: very often users already know why a dependency cannot be found, e.g. they never installed it, and too much output runs the risk of drowning out the core details.

**Comment by ilayn** [comment]
> very often users already know why a dependency cannot be found, e.g. they never installed it, and too much output runs the risk of drowning out the core details.

For us this is exactly the opposite. Well at least in the setup.cfg way of doing it. Dependencies are very foreign to Python users because of the lack of build experience. This is one of the extremely annoying details about built code. It assumes you already know what to do and the options are hidden behind some obscure flag combination. 

Meson might want to skip it but if possible I would like to work on the error messages. This is practically how Rust got to the point where it is now. By being very explicit about the cause and just nudging the user about the solution. Unfortunately compiler related tools are by legacy, by culture or whatever the cause is very terse and cryptic.

**Comment by rgommers** [comment]
> For us this is exactly the opposite. Well at least in the setup.cfg way of doing it.

I'm not sure I agree with that. One of the major issues with distutils builds is that our build log is 5000+ lines long, so users and maintainers often miss clear errors because they're drowned out by other output. I'm much happier with what Meson does.

**Comment by ilayn** [comment]
For example this is a succesful find; 

```bash
  libraries openblas not found in ['C:\\Users\\ilhan\\AppData\\Local\\Programs\\Python\\Python39\\lib', 'C:\\', 'C:\\Users\\ilhan\\AppData\\Local\\Programs\\Python\\Python39\\libs']
Library openblas was not found. Ignoring
  FOUND:
    library_dirs = ['C:\\Users\\ilhan\\Documents\\GitHub\\scipy\\build\\openblas_info']
    libraries = ['openblas_info']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]

  FOUND:
    library_dirs = ['C:\\Users\\ilhan\\Documents\\GitHub\\scipy\\build\\openblas_info']
    libraries = ['openblas_info']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]
```

To me this makes no sense what is happening here.

**Comment by eli-schwartz** [comment]
And in Meson, a successful find looks like this:
```
Found pkg-config: /usr/bin/pkg-config (1.8.0)
Run-time dependency openblas found: YES 0.3.19
```

An unsuccessful one is slightly longer:
```
Found pkg-config: /usr/bin/pkg-config (1.8.0)
Found CMake: /usr/bin/cmake (3.22.2)
Run-time dependency openblas found: NO (tried pkgconfig and cmake)

meson.build:5:0: ERROR: Dependency "openblas" not found, tried pkgconfig and cmake

A full log can be found at /tmp/builddir/meson-logs/meson-log.txt
```

The final message which triggered an error gets repeated with slightly different wording, then it tells you where to find the verbose logs on what/why (and the verbose log file does indeed tell you, I think, everything you could want to know about what happened).

I would welcome discussion on improving Meson's output. But my personal feeling is that it's already pretty darn good. :)

**Comment by rgommers** [comment]
I did some testing. First tried this: A test with Netlib BLAS/LAPACK from conda:
```
$ conda activate scipy-dev  # our regular dev env
$ conda install blas lapack blas-devel  # install Netlib BLAS/LAPACK
$ git clean -xdf
$ meson setup build -Dblas=blas -Dlapack=lapack --prefix=$PWD/build-install
$ python dev.py  # run test suite (passes)
```
This actually still uses OpenBLAS because of how conda-forge deals with multi-BLAS switching (see [docs](https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=blas#switching-blas-implementation)).


Second try:
```
$ git clean -xdf
$ conda env create -f environment.yml
$ conda install meson ninja blas-devel "libblas=*=*netlib"   # this actually selects the libraries we want
$ meson setup build -Dblas=blas -Dlapack=lapack --prefix=$PWD/build-install
$ python dev.py
....
E   ImportError: /home/rgommers/code/scipy/build-install/lib/python3.10/site-packages/scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so: undefined symbol: cspmv_
```
So that reproduces the error @ev-br was seeing.

Netlib BLAS works on the conda-forge builds though, so unclear what is different. It's built there with a g77 ABI (see https://github.com/conda-forge/scipy-feedstock/blob/master/recipe/build.sh#L5), but that should only affect the few dummy symbols we use.

Looking at the installed Netlib `libblas`:
```
$ readelf -s libblas.so | grep spmv
    16: 00000000000070d0  2375 FUNC    GLOBAL DEFAULT   11 sspmv_
    40: 0000000000018ba0  2255 FUNC    GLOBAL DEFAULT   11 dspmv_
    58: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS sspmv.f
    93: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS dspmv.f
   205: 00000000000070d0  2375 FUNC    GLOBAL DEFAULT   11 sspmv_
   323: 0000000000018ba0  2255 FUNC    GLOBAL DEFAULT   11 dspmv_
```
So no complex flavors of `spmv`? Not quite - what seems to be happening here is that the float/double versions are part of BLAS and the others are part of LAPACK, at least according to our Cython API signatures:
```
scipy/_build_utils/src/npy_cblas_base.h
278:void BLASNAME(cblas_sspmv)(const enum CBLAS_ORDER order, const enum CBLAS_UPLO Uplo,
307:void BLASNAME(cblas_dspmv)(const enum CBLAS_ORDER order, const enum CBLAS_UPLO Uplo,

scipy/linalg/fblas_l2.pyf.src
10:!   spmv, hpmv, hbmv, sbmv, hpr2, spr2, gbmv, trsv, tbmv, tpmv, tbsv, tpsv

scipy/linalg/cython_lapack_signatures.txt
261:void cspmv(char *uplo, int *n, c *alpha, c *ap, c *x, int *incx, c *beta, c *y, int *incy)
1405:void zspmv(char *uplo, int *n, z *alpha, z *ap, z *x, int *incx, z *beta, z *y, int *incy)

scipy/linalg/tests/test_blas.py
519:    def test_spmv_hpmv(self):
539:                func, = get_blas_funcs(('spmv',), dtype=dtype)

scipy/linalg/cython_blas_signatures.txt
56:void dspmv(char *uplo, int *n, d *alpha, d *ap, d *x, int *incx, d *beta, d *y, int *incy)
99:void sspmv(char *uplo, int *n, s *alpha, s *ap, s *x, int *incx, s *beta, s *y, int *incy)
```

So are they in LAPACK? Yes:
```
$ readelf -s liblapack.so | grep cspmv
   479: 000000000043b000  5722 FUNC    GLOBAL DEFAULT   11 cspmv_
  1344: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS cspmv.f
  3024: 000000000043b000  5722 FUNC    GLOBAL DEFAULT   11 cspmv_
```
Checking the official docs at http://www.netlib.org/blas/#_level_2 shows that indeed only `sspmv` and `dspmv` are listed. And `cspmv` and `zspmv` are part of LAPACK indeed: http://www.netlib.org/lapack/explore-html/dd/d38/cspmv_8f.html

We never see this for OpenBLAS, MKL or ATLAS because `libblas` and `liblapack` aren't separated. However for Netlib we do see it (unclear to me why we don't have this issue on conda-forge right now). 

tl;dr it's a pre-existing bug in our `fblas` f2py signature files.

**Comment by rgommers** [comment]
Ah looking at a build log from https://github.com/conda-forge/scipy-feedstock/pull/199 for a Netlib job, I see:
```
-llapack -lblas -llapack -lblas -lcblas -lblas -lcblas -lblas -lgfortran -o build/lib.linux-x86_64-3.9/scipy/linalg/_fblas.cpython-39-x86_64-linux-gnu.so
```
While with Meson we see:
```ninja
build scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so: fortran_LINKER scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so.p/meson-generated_..__fblasmodule.c.o scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so.p/meson-generated_..__fblas-f2pywrappers.f.o scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so.p/_home_rgommers_anaconda3_envs_scipy-dev_lib_python3.10_site-packages_numpy_core_include_.._.._f2py_src_fortranobject.c.o scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so.p/..__build_utils_src_wrap_dummy_g77_abi.f.o | /home/rgommers/anaconda3/envs/scipy-dev/lib/libblas.so
 LINK_ARGS = -Lers/anaconda3/envs/scipy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/9.4.0/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../lib/gcc/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/9.4.0/../../../../x86_64-conda-linux-gnu/lib/x86_64-conda-linux-gnu/9.4.0/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/9.4.0/../../../../x86_64-conda-linux-gnu/lib/../lib/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/9.4.0/../../../x86_64-conda-linux-gnu/9.4.0/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/9.4.0/../../../../lib/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../x86_64-conda-linux-gnu/sysroot/lib/x86_64-conda-linux-gnu/9.4.0/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../x86_64-conda-linux-gnu/sysroot/lib/../lib/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../x86_64-conda-linux-gnu/sysroot/usr/lib/x86_64-conda-linux-gnu/9.4.0/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../x86_64-conda-linux-gnu/sysroot/usr/lib/../lib/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/9.4.0/../../../../x86_64-conda-linux-gnu/lib/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/9.4.0/../../../ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../x86_64-conda-linux-gnu/sysroot/lib/ -L/home/rgommers/anaconda3/envs/scipy-dev/bin/../x86_64-conda-linux-gnu/sysroot/usr/lib/ -L/home/rgommers/anaconda3/envs/scipy-dev/lib -Wl,--as-needed -Wl,--allow-shlib-undefined -shared -fPIC -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/home/rgommers/anaconda3/envs/scipy-dev/lib -Wl,-rpath-link,/home/rgommers/anaconda3/envs/scipy-dev/lib -Wl,--start-group /home/rgommers/anaconda3/envs/scipy-dev/lib/libblas.so -lgfortran -lm -Wl,--end-group
```
(long lines, but tl;dr it is only linking explicitly `libblas.so`)

A Meson build:
```
$ readelf -d build/scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libblas.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```
A distutils build will link LAPACK - from `linalg/setup.py`:
```python
    # fblas:
    sources = ['fblas.pyf.src']
    sources += get_g77_abi_wrappers(lapack_opt)
    depends = ['fblas_l?.pyf.src']

    config.add_extension('_fblas',
                         sources=sources,
                         depends=depends,
                         extra_info=lapack_opt
                         )
```
So we're on purpose letting `_fblas.so` depend on LAPACK. Not sure why.

**Comment by rgommers** [comment]
I went with the option of linking `liblapack` into the `_fblas` and `cython_blas` extensions - we've been doing this as far back as 2007 at least for `_fblas` and from the start for `cython_blas`, so it doesn't make sense to spend a lot of effort now to more cleanly separate the wrappers.

This now works for me with Netlib BLAS/LAPACK with:
```
$ meson setup build -Dblas=blas -Dlapack=lapack --prefix=$PWD/build-install
$ python dev.py
```

**Comment by ev-br** [comment]
Confirmed that the current state of the PR (67704de) builds fine on the ubuntu focal / netlib with pip, test pass locally. Thanks @rgommers @Smit-create !

**Comment by ilayn** [comment]
Same with Windows. I can do separate MKL install too. So I think this solves it.

But this is yet another example how difficult it is to debug these issues, hence I repeat my case (@eli-schwartz)

**Comment by rgommers** [comment]
Thanks for testing @ev-br and @ilayn, that's all good news. In it goes then.

> I can do separate MKL install too

@ilayn can you give some more details or instructions for reproducing that? My understanding is that pkg-config files for MKL are not great, so it'd be good to understand what does work here.

> But this is yet another example how difficult it is to debug these issues,

Any change in logging wouldn't have helped with the hairy part of this I think. What did help, and is a major upgrade, is that we can inspect the compile and link flags without actually running the build (using `meson introspect` and/or reading the generated `build.ninja` file).

**Comment by eli-schwartz** [comment]
@ilayn I'm not entirely clear on what kind of logging you'd like to have here. Both blas and lapack were found -- and correctly! -- in all cases and the issue was solely in whether the resolved lapack flags were added to build targets that needed it.

At which stage of the configuration would you like to see better logging? The lapack dependency lookup? The `_fblas` build target creation? Which specific details would have helped in this case, and which details do you believe would help as a general thing?

In summary: you've stated you believe the logging could be "better". I can't do much with that information, and am interested to hear an answer to the question "better in what way?"

**Comment by ilayn** [comment]
@eli-schwartz As I mentioned, the build tools unanimously assumes that if you are building something you know what you are doing. That is just false. But since every tool does that, anytime this discussion pops up we start the same trail of discussion. 

What the tool is doing, and meson is no exception, is just printing every possible flag and that's pretty much it. And error message is just that. At this step there is an error. Good luck with that. No context, no "I'm trying this now", no "I found this", no "I'm linking this with this tool with this path". 

I am not saying "it should be just better" that's you putting words in my mouth. Just look above and tell me that's a debugging sequence that a relatively average user can do without too much guidance. 

In summary; I am not demanding it to be better. It is just feedback, do whatever you want with it.

**Comment by rgommers** [comment]
> Just look above and tell me that's a debugging sequence that a relatively average user can do without too much guidance.

They can't, and I bet a good fraction of maintainers would have trouble with this too:) This is not an issue that users should be debugging though - if the bug is not user error but an actual bug in our own build config, the best we can do is ask users for logs and a good reproducer.

**Comment by eli-schwartz** [comment]
> I am not saying "it should be just better" that's you putting words in my mouth.

Okay, I won't put any words in your mouth.

> But this is yet another example how difficult it is to debug these issues, hence I repeat my case

This "another example" doesn't map in any way to the previous case, so I don't know what you think *this example* could do better.

Perhaps it should be just better, but not better in any *specific manner*? 

> And error message is just that. At this step there is an error. Good luck with that. No context, no "I'm trying this now", no "I found this", no "I'm linking this with this tool with this path". Just look above and tell me that's a debugging sequence that a relatively average user can do without too much guidance.

But wait! Errors? What errors? This "another example" is a case where there were no errors at all! Meson succeeded, following the instructions written into `meson.build`. Ninja also built successfully, and produced a full SciPy install...

... which happened to be missing symbols as an ImportError, but only at runtime when you try to run a program using the built SciPy.

How is Meson supposed to interject and report better context on the error?

Meson could report better context as part of some verbose configuration logging during a normal build, optimizing for the case that errors occur after installation. But that won't be attached to the error itself. Also, you'd end up with hundreds or thousands of lines of interesting context about stuff you're not interested in right now (but may be interested in some other day).

Meson could do what it always does when building libraries, and pass the right compiler options to cause undefined symbols to be a build error, which moves the error message one step closer. Except this is a shared_module, it has lots of undefined symbols from libpython, which are injected by the interpreter, so even the compiler doesn't know whether lapack symbols being missing are an error.

Errors that only occur at runtime, long after configuration, building, and installing are over and done with, are a worst-case scenario to debug.

...

And that is why I don't know what Meson could be doing to make this easier to debug. You *have* to work backwards from "X has failed at runtime, how can I ask the build system for more information about X".

Maybe if there was an option to re-run `meson` with some kind of `--debug-build-target _fblas` option that would rerun everything, but this time, print lots of interesting details related to the one thing you're interested in.

Although as @rgommers pointed out, you can already run `meson introspect` or scrape the json yourself, and get information about the targets and how they were built. Which turned out to be the critical information here.
